crates/storage/src/iter/changes_iterator.rs

```diff
@@ -88,16 +141,62 @@ where
         // because we have to filter out the keys that were removed, which are
         // marked as `WriteOperation::Remove` in the value
         // copied as-is from the above function, but only to return keys
-        if let Some(tree) = self.changes.get(&column.id()) {
-            crate::iter::iterator(tree, prefix, start, direction)
-                .filter_map(|(key, value)| match value {
-                    WriteOperation::Insert(_) => Some(key.clone().into()),
-                    WriteOperation::Remove => None,
-                })
-                .map(Ok)
-                .into_boxed()
-        } else {
-            core::iter::empty().into_boxed()
+        match self.changes {
+            StorageChanges::Changes(changes) => get_insert_keys_from_changes(
+                changes,
+                column.id(),
+                prefix,
+                start,
+                direction,
+            ),
+            StorageChanges::ChangesList(changes_list) => {
+                // We have to clone the prefix and start, because we need to pass them to the iterator
+                // if someone finds a solution without making it a vec, feel free to contribute :)
+                let column = column.id();
+                let prefix = prefix.map(|prefix| prefix.to_vec());
+                let start = start.map(|start| start.to_vec());
+                changes_list
+                    .iter()
+                    .flat_map(move |changes| {
+                        get_insert_keys_from_changes(
+                            changes,
+                            column,
+                            prefix.as_deref(),
+                            start.as_deref(),
+                            direction,
+                        )
+                    })
+                    .into_boxed()
+            }
```

Comment:
IDK if this would be better or worse, but it avoids the `to_vec`s :P
```suggestion
            StorageChanges::ChangesList(changes_list) => {
                let column = column.id();
                changes_list
                    .iter()
                    .flat_map(move |changes| {
                        get_insert_keys_from_changes(
                            changes, column, prefix, start, direction,
                        )
                    })
                    .collect::<Vec<_>>()
                    .into_iter()
                    .into_boxed()
            }
```

Reply:
mehhh, i don't really like that because start and prefix are pretty small in most of the cases where the number of elements iterated on could be pretty big so iterating twice and creating a heap allocation with all elements will be more costly than 2 little allocation